service/tbc/zkindexer_test.go

```diff
@@ -0,0 +1,200 @@
+// Copyright (c) 2025 Hemi Labs, Inc.
+// Use of this source code is governed by the MIT License,
+// which can be found in the LICENSE file.
+
+package tbc
+
+import (
+	"bytes"
+	"context"
+	"crypto/rand"
+	"encoding/binary"
+	"fmt"
+	"testing"
+
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"github.com/btcsuite/btcd/wire"
+
+	"github.com/hemilabs/heminetwork/v2/database/tbcd"
+	"github.com/hemilabs/heminetwork/v2/database/tbcd/level"
+	"github.com/hemilabs/heminetwork/v2/testutil"
+)
+
+func TestZKEncodeRetrieve(t *testing.T) {
+	ctx, cancel := context.WithCancel(t.Context())
+	defer cancel()
+
+	home := t.TempDir()
+	cfg, err := level.NewConfig("localnet", home,
+		"", "")
+	if err != nil {
+		t.Fatal(err)
+	}
+	db, err := level.New(ctx, cfg)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Create all types of indexes using encoders
+	cache := make(map[tbcd.ZKIndexKey][]byte, 5)
+	index := 99
+	blockHash := randomHash()
+	txId := randomHash()
+	prevHash := randomHash()
+	var balance uint64 = 10
+
+	// ScriptHash
+	shbytes := testutil.FillBytes(fmt.Sprintf("scripthash%d", index), len(tbcd.ScriptHash{}))
+	sh := tbcd.NewScriptHashFromScript(shbytes)
+	var v [8]byte
+	binary.BigEndian.PutUint64(v[:], balance)
+	cache[tbcd.ZKIndexKey(sh[:])] = v[:]
+
+	// SpentOut
+	spo := tbcd.NewSpentOutput(chainhash.Hash(sh), uint32(index),
+		*blockHash, *txId, *prevHash, uint32(index-1), 0)
+	cache[tbcd.ZKIndexKey(spo[:])] = nil
+
+	// SpendingOut
+	sok := tbcd.NewSpendingOutpointKey(*prevHash,
+		uint32(index), *blockHash, uint32(index-1))
+	cache[tbcd.ZKIndexKey(sok[:])] = tbcd.NewSpendingOutpointValueSlice(*txId, uint32(index))
+
+	// SpendableOut
+	so := tbcd.NewSpendableOutput(chainhash.Hash(sh), uint32(index),
+		*blockHash, *txId, 0)
+	cache[tbcd.ZKIndexKey(so[:])] = nil
+
+	// Outpoint
+	op := tbcd.NewOutpoint(*txId, 0)
+	txOut := wire.TxOut{Value: 10, PkScript: txId[:]}
+	cache[tbcd.ZKIndexKey(op[:])] = tbcd.NewTxOut(&txOut)
+
+	// Insert into DB
+
+	err = db.BlockZKUpdate(ctx, 1, cache, chainhash.Hash{})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Retrieve values by encoded key and assert values
+
+	// Outpoint
+	rv, rid, err := db.ZKValueAndScriptByOutpoint(ctx, op)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if balance != rv {
+		t.Fatalf("expected %v, got %v", balance, rv)
+	}
+	if !bytes.Equal(txId[:], rid) {
+		t.Fatalf("expected %x, got %x", txId, rid)
+	}
+
+	// ScriptHash
+	rv, err = db.ZKBalanceByScriptHash(ctx, sh)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if balance != rv {
+		t.Fatalf("expected %v, got %v", balance, rv)
+	}
+
+	// SpentOut
+	out, err := db.ZKSpentOutputs(ctx, sh)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if len(out) != 1 {
+		t.Fatalf("expected 1 spent output, got %d", len(out))
```

Comment:
Go generally prefers "got before want", and identifying the function in tests, per:
https://go.dev/wiki/TestComments#got-before-want
https://go.dev/wiki/TestComments#identify-the-function

This change makes the failure easier to debug and a lot more descriptive, whilst being more idiomatic:
```suggestion
		t.Fatalf("ZKSpentOutputs out len = %d, want 1", len(out))
```

Go avoids "expected, got" or "expected, actual", and opts for "got X, want Z" or better `MyFunc(%q) = %v, want %v`

Also, it is generally preferred to use `t.Error` instead of `t.Fatal` when possible (as long as the test can actually continue) and/or using `t.Run` to create sub-tests to avoid failing at the first error, making it easier to figure out what broke.